### DISCUSSION TITLE
How to find and install binaries (.dll or .so) for the target

### DISCUSSION THREAD

**Discussion by jfulem** [xmake-io/xmake#5557] [Q&A] [answered]
Please, is there any good way how to install or copy binaries (.dll or .so files) to the build directory? These binaries come from:
 `add_requires()` and `add_packages()`

Or is it possible to get the target dependencies pointing to `xmake/xrepo package`s (installation)?

Thank you very much for your help.

**Comment by waruqi** [comment, marked as answer]
run `xmake install -o build` will install them.

if you debug program, you can run `xmake run`. it will add PATH to load dll automatically.

**Reply by jfulem** [reply]
Great! This works, thank you. 

Can I have one more question? I'm working with OpenCV 4.10.0 and another package called depthai-core. I'm able to install this package in my xmake.lua via import("package.tools.cmake").install(), but inside the depthai-core cmake files there is a check whether OpenCV is installed or not and it looks the OpenCV is not found even though I used add_deps() in the package in xmake.lua. 

The depthai-core suggestion,in the case of OpenCV is not found but it is installed, is to set the OpenCV_DIR environment variable, but I'm not sure how this can be done so it's pointing to xmake installed package. 

Any suggestion how fix this would be great. Thank you.

**Reply by waruqi** [reply]
like this patch. https://github.com/xmake-io/xmake-repo/pull/4938

**Reply by jfulem** [reply]
@waruqi, thank you for the previous answers. I still have some small issues when using the cmake package. One dynamic library is not copied during the installation. It never worked in cmake project. 

So is there any way to get the installation directory of a package ("depthai-core") added to the target as a package. I would just add one specific .dll to the installation in _**before_install**_ step. I can't find any way to get the "depthai-core" package install or linkdir where the specific .dll is located. Thank you.

```
target("camera")
    -- building executable
    set_kind("binary")

    -- add required packages
    add_packages("depthai-core")
    
    before_install(function (target)
        -- get the "depthai-core" installation directory
        add_installfiles("", {prefixdir = "bin"})
    end)
    ```

**Reply by star-hengxing** [reply]
`target:dep("x"):installdir()`

**Reply by jfulem** [reply]
> target:dep("x"):installdir()

This does not work. Nothing found by target:dep()

I'm not sure why one of the dlls is not copied during the installation. My executable loads A.dll and this loads B.dll, but this one is not copied. Not sure how to resolve this :-(.

**Reply by waruqi** [reply]
you can use `print` to debug here.

https://github.com/xmake-io/xmake/blob/9563841bfc304874cc9d308b1e29bd6cde76631a/xmake/modules/target/action/install/main.lua#L48-L72

**Reply by star-hengxing** [reply]
My mistake, package use `target:pkg("zlib")`
